File: libs/grpc/clients/tokens.ts
import { ChannelCredentials, createChannel, createClient } from 'nice-grpc'

import { CLIENT_OPTIONS, createClientAddr, waitReady } from './helper'
import { PORTS } from '../constants/constants.js'
import { TokensDefinition } from '../tokens/tokens'

import type { TokensClient } from '../tokens/tokens'

export async function createTokens(env: string): Promise<TokensClient> {
	const channel = createChannel(
		createClientAddr(env, 'tokens', PORTS.TOKENS_SERVER_PORT),
		ChannelCredentials.createInsecure(),
		CLIENT_OPTIONS,
	)

	await waitReady(channel)

	const client = createClient(TokensDefinition, channel)

	return client as any
}

File: libs/grpc/clients/tokens.go
package clients

import (
	"log"

	"github.com/twirapp/twir/libs/grpc/constants"
	"github.com/twirapp/twir/libs/grpc/tokens"

	"google.golang.org/grpc"
)

func NewTokens(env string) tokens.TokensClient {
	serverAddress := createClientAddr(env, "tokens", constants.TOKENS_SERVER_PORT)

	conn, err := grpc.Dial(
		serverAddress,
		defaultClientsOptions...,
	)
	if err != nil {
		log.Fatalf("did not connect: %v", err)
	}
	c := tokens.NewTokensClient(conn)

	return c
}
